File: LeetCode/medium/Create_Binary_Tree_From_Descriptions.cxx
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        unordered_map<int, TreeNode*> nodes;
        unordered_map<int, bool> is_root;

        for (auto& desc : descriptions) {
            if (nodes.count(desc[0]) == 0)
                nodes[desc[0]] = new TreeNode(desc[0]);
            if (nodes.count(desc[1]) == 0)
                nodes[desc[1]] = new TreeNode(desc[1]);
            is_root[desc[0]] = true;
            is_root[desc[1]] = true;
        }
        for (auto& desc : descriptions) {
            if (desc[2])
                nodes[desc[0]]->left = nodes[desc[1]];
            else
                nodes[desc[0]]->right = nodes[desc[1]];
            is_root[desc[1]] = false;
        }
        for (auto b : nodes) {
            if (is_root[b.first])
                return b.second;
        }
        return nullptr;
    }
};

File: LeetCode/medium/Create_Binary_Tree_From_Descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        nodes = {}
        children = set()
        for parent, child, side in descriptions:
            if parent not in nodes.keys():
                nodes[parent] = TreeNode(parent)
            if child not in nodes.keys():
                nodes[child] = TreeNode(child)
            if side:
                nodes[parent].left = nodes[child]
            else:
                nodes[parent].right = nodes[child]
            children.add(child)

        for n in nodes.keys():
            if n not in children:
                return nodes[n]

        
